chiselFrontend/src/main/scala/chisel3/internal/Builder.scala

```diff
@@ -142,12 +151,20 @@ private[chisel3] trait HasId extends InstanceId {
     this.getClass.getMethods.sortWith(_.getName < _.getName).filter(isPublicVal(_))
   }
 }
+/** Holds the implementation of toNamed for Data and MemBase */
+private[chisel3] trait NamedComponent extends HasId {
+  /** Returns a FIRRTL ComponentName that references this object
+    * @note Should not be called until circuit elaboration is complete
```

Comment:
Perhaps not a now thing, but can we relax when this can be called? For example, it should be possible to get the name of any bound wire, assuming it's properly named (instead of T_, which is kind of meaningless anyways) - though we might need to ensure that these names can't be mangled in the future. Going up the module chain also presents difficulties in this case, you never know when an enclosing module might not have a proper name.

Reply:
I think making it possible to eagerly access the names of things is desirable but it's a non-trivial undertaking. We would need to ensure that querying the name of something does not impact the result. For example:
```scala
class NamingTop extends RawModule {
  val w = Wire(UInt(8.W)).suggestName("foo")
  w.getName // ???
  val foo = IO(Input(UInt(8.W)))
  w := foo
}
```
Currently, `w` gets the name `foo_1` because ports always win. We would need to be sure that calling `getName` did not accidentally change this.

Doing this with reflection also sounds... dangerous. Perhaps we could make `@chiselName` disable the reflection-based naming and make eager-naming a feature of using `@chiselName`?

Reply:
I still can't say I'm a fan of an API that depends so much on mutable state where access can be unsafe. Are there other possible APIs, like ChiselAnnotation.toFirrtl taking in a object that can remap a component to a name?

Reply:
What is the difference between putting the code here in `NamedComponent` vs. in `ChiselAnnotation`? In either case, getting a Firrtl `Named` from a Chisel object relies on mutable state in that object that isn't set until the end of elaboration. This PR is not changing anything about that aspect of Chisel's implementation.